XDMA/linux-kernel/xdma/alinx_arch.c

```diff
@@ -74,6 +74,16 @@ timestamp_t alinx_read_tx_timestamp(struct pci_dev* pdev, int tx_id) {
         }
 }
 
+u32 alinx_get_tx_packets(struct pci_dev *pdev) {
+        struct xdma_dev* xdev = xdev_find_by_pdev(pdev);
+        return read32(xdev->bar[0] + REG_TX_PACKETS);
+}
+
+u32 alinx_get_tx_drop_packets(struct pci_dev *pdev) {
+        struct xdma_dev* xdev = xdev_find_by_pdev(pdev);
+        return read32(xdev->bar[0] + REG_TX_DROP_PACKETS);
```

Comment:
This register is cleared at read. So this could side effect by calling it from multiple location.
Read and add up to pdev->something and return total value of it.

Reply:
And buffer tracker can use their own static variable to track how many packets are sent/dropped

Reply:
I added `tx_count` and `tx_drop_count` variables to `xdma_private`. Also, I made the buffer tracker to track the total number of sent/dropped packets. That makes tracking the buffer's state require two variables, so I added `struct buffer_tracker` and put those two variables there. `last_tx_count` counts both sent and dropped packets, since I thought this does not need to consider whether the packets were sent or not.

https://github.com/tsnlab/2023-krri-tci-xdma/blob/22a9559481a8ed5d0f9f46903ece0d4ebb37c628/XDMA/linux-kernel/xdma/alinx_arch.h#L117-L120